File: ios/Sources/Core/XTDebug.m
//
//  XTRDebug.m
//  XTSample
//
//  Created by 崔明辉 on 2017/9/19.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTDebug.h"
#import "XTContext.h"
#import "XTUIContext.h"
#import <SocketRocket/SRWebSocket.h>

@interface XTDebug ()<SRWebSocketDelegate>

@property (nonatomic, strong) NSURL *sourceURL;
@property (nonatomic, strong) SRWebSocket *socket;
@property (nonatomic, assign) BOOL breakpointLocking;
@property (nonatomic, assign) BOOL breakpointStepping;

@end

@implementation XTDebug

+ (NSString *)name {
    return @"_XTDebug";
}

+ (XTDebug *)sharedDebugger {
    static XTDebug *debuger;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        debuger = [XTDebug new];
    });
    return debuger;
}

+ (void)debugWithIP:(NSString *)IP port:(NSInteger)port navigationController:(UINavigationController *)navigationController {
    [[XTDebug sharedDebugger] connectWithIP:IP port:port];
    [UIApplication sharedApplication].networkActivityIndicatorVisible = YES;
}

+ (void)xtr_break:(NSString *)bpIdentifier T:(NSString *)T S:(NSString *)S {
    [[XTDebug sharedDebugger] handleBreak:bpIdentifier T:T S:S];
}

+ (void)xtr_wait {
    [NSThread sleepForTimeInterval:0.1];
}

+ (BOOL)xtr_locking {
    return [XTDebug sharedDebugger].breakpointLocking;
}

+ (BOOL)xtr_stepping {
    return [XTDebug sharedDebugger].breakpointStepping;
}

- (void)connectWithIP:(NSString *)IP port:(NSInteger)port {
    NSString *URLString = [NSString stringWithFormat:@"ws://%@:%ld", IP, (long)port];
    if (self.socket != nil) {
        [self.socket close];
    }
    self.socket = [[SRWebSocket alloc] initWithURLRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:URLString]]];
    self.socket.delegate = self;
    [self.socket setDelegateOperationQueue:[NSOperationQueue new]];
    [self.socket open];
}

- (void)sendLog:(NSString *)string isEval:(BOOL)isEval {
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        NSString *currentBpIdentifier = [self.debugContext evaluateScript:@"window.XTDebug.currentBpIdentifier"].toString;
        NSDictionary *obj = @{
                              @"type": @"console.log",
                              @"payload": ([[string dataUsingEncoding:NSUTF8StringEncoding] base64EncodedStringWithOptions:kNilOptions] ?: @""),
                              @"bpIdentifier": isEval ? @"console" : currentBpIdentifier,
                              };
        NSData *data = [NSJSONSerialization dataWithJSONObject:obj options:kNilOptions error:NULL];
        [self.socket send:[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
    }];
}

#pragma mark - Breakpoint

- (void)sendBreak:(NSString *)bpIdentifier T:(NSString *)T S:(NSString *)S {
    NSDictionary *obj = @{
                          @"type": @"break",
                          @"bpIdentifier": bpIdentifier ?: @"",
                          @"this": T ?: @"{}",
                          @"scope": S ?: @"{}",
                          };
    NSData *data = [NSJSONSerialization dataWithJSONObject:obj options:kNilOptions error:NULL];
    [self.socket send:[[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding]];
}

- (void)handleBreak:(NSString *)bpIdentifier T:(NSString *)T S:(NSString *)S {
    [self sendBreak:bpIdentifier T:T S:S];
    self.breakpointLocking = YES;
}

#pragma mark - WebSocket Deleagte

- (void)webSocketDidOpen:(SRWebSocket *)webSocket {
    [webSocket send:@"Hello, World!"];
}

- (void)webSocket:(SRWebSocket *)webSocket didReceiveMessage:(id)message {
    if ([message isKindOfClass:[NSString class]]) {
        NSDictionary *obj = [NSJSONSerialization JSONObjectWithData:[message dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:NULL];
        if ([obj isKindOfClass:[NSDictionary class]] && [obj[@"action"] isKindOfClass:[NSString class]]) {
            if ([obj[@"action"] isEqualToString:@"reload"]) {
                [self handleReload:obj];
            }
            else if ([obj[@"action"] isEqualToString:@"clearBreakPoint"]) {
                [self handleClearBreakPoint:obj];
            }
            else if ([obj[@"action"] isEqualToString:@"clearBreakPoints"]) {
                [self handleClearBreakPoints:obj];
            }
            else if ([obj[@"action"] isEqualToString:@"setBreakPoint"]) {
                [self handleSetBreakPoint:obj];
            }
            else if ([obj[@"action"] isEqualToString:@"continue"]) {
                [self handleContinue];
            }
            else if ([obj[@"action"] isEqualToString:@"stop"]) {
                [self handleStop];
            }
            else if ([obj[@"action"] isEqualToString:@"step"]) {
                [self handleStep];
            }
            else if ([obj[@"action"] isEqualToString:@"eval"]) {
                [self handleEval:obj];
            }
        }
    }
}

- (void)webSocket:(SRWebSocket *)webSocket didCloseWithCode:(NSInteger)code reason:(NSString *)reason wasClean:(BOOL)wasClean {
    self.breakpointLocking = NO;
    self.breakpointStepping = NO;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.socket = [[SRWebSocket alloc] initWithURL:[webSocket url]];
        self.socket.delegate = self;
        [self.socket open];
    });
}

- (void)webSocket:(SRWebSocket *)webSocket didFailWithError:(NSError *)error {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.socket = [[SRWebSocket alloc] initWithURL:[webSocket url]];
        self.socket.delegate = self;
        [self.socket open];
    });
}

#pragma mark - Message Handler

- (void)handleReload:(NSDictionary *)obj {
    if ([obj[@"source"] isKindOfClass:[NSString class]]) {
        NSData *sourceData = [[NSData alloc] initWithBase64EncodedString:obj[@"source"] options:kNilOptions];
        if (sourceData != nil) {
            NSString *tmpPath = [NSString stringWithFormat:@"%@/%u.min.js", NSTemporaryDirectory(), arc4random()];
            [sourceData writeToFile:tmpPath atomically:YES];
            self.sourceURL = [NSURL fileURLWithPath:tmpPath];
            self.breakpointStepping = NO;
            self.breakpointLocking = NO;
            [[NSOperationQueue mainQueue] addOperationWithBlock:^{
                [self.delegate debuggerDidReload];
            }];
        }
    }
}

- (void)handleClearBreakPoint:(NSDictionary *)obj {
    NSString *bpIdentifier = [NSString stringWithFormat:@"%@:%@", obj[@"path"], obj[@"line"]];
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        JSValue *obj = [self.debugContext evaluateScript:@"window.XTDebug"];
        [obj invokeMethod:@"clearBreakpoint" withArguments:@[bpIdentifier]];
    }];
}

- (void)handleClearBreakPoints:(NSDictionary *)obj {
    NSString *path = obj[@"path"] ?: @"";
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        JSValue *obj = [self.debugContext evaluateScript:@"window.XTDebug"];
        [obj invokeMethod:@"clearBreakpoints" withArguments:@[path]];
    }];
}

- (void)handleSetBreakPoint:(NSDictionary *)obj {
    NSString *bpIdentifier = [NSString stringWithFormat:@"%@:%@", obj[@"path"], obj[@"line"]];
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        JSValue *obj = [self.debugContext evaluateScript:@"window.XTDebug"];
        [obj invokeMethod:@"setBreakpoint" withArguments:@[bpIdentifier]];
    }];
}

- (void)handleContinue {
    self.breakpointStepping = NO;
    self.breakpointLocking = NO;
}

- (void)handleStep {
    self.breakpointStepping = YES;
    self.breakpointLocking = NO;
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        self.breakpointStepping = NO;
    });
}

- (void)handleStop {
    self.breakpointStepping = NO;
    self.breakpointLocking = NO;
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        [self.delegate debuggerDidTerminal];
    }];
}

- (void)handleEval:(NSDictionary *)obj {
    if (![obj[@"expression"] isKindOfClass:[NSString class]]) {
        return;
    }
    [[NSOperationQueue mainQueue] addOperationWithBlock:^{
        NSString *result = [self.delegate debuggerEval:obj[@"expression"]];
        if ([result isKindOfClass:[NSString class]]) {
            [self sendLog:result isEval:YES];
        }
    }];
}

@end

File: android/core/src/main/java/com/opensource/xt/core/XTDebug.kt
package com.opensource.xt.core
import android.content.Context
import android.net.Uri
import android.os.Handler
import android.util.Base64
import com.eclipsesource.v8.V8Object
import org.java_websocket.WebSocket
import org.java_websocket.client.WebSocketClient
import org.java_websocket.handshake.ServerHandshake
import org.json.JSONObject
import java.io.File
import java.lang.Exception
import java.net.URI

/**
 * Created by cuiminghui on 2018/2/1.
 */

interface XTDebugDelegate {

    fun debuggerDidTerminal()
    fun debuggerDidReload()
    fun debuggerEval(code: String): String

}

class XTDebug {

    var sourceURL: String? = null
    var delegate: XTDebugDelegate? = null
    var xtContext: XTContext? = null
    private var context: Context? = null
    private var handler = Handler()
    private var socket: WebSocket? = null
        set(value) {
            field?.let { it.close() }
            field = value
        }
    private var activeBreakpoints: MutableSet<String> = mutableSetOf()
    private var currentBpIdentifier: String? = null
    private var breakpointLocking = false
    private var breakpointStepping = false

    private fun connectWithIP(IP: String, port: String, context: Context) {
        this.context = context
        this.socket = object: WebSocketClient(URI.create("ws://$IP:$port/")) {

            init {
                connect()
            }

            override fun onOpen(data: ServerHandshake?) {
                send("Hello, World!")
            }

            override fun onClose(code: Int, reason: String?, remote: Boolean) {
                if (remote) {
                    handleStop()
                    handler.postDelayed({
                        connectWithIP(IP, port, context)
                    }, 1000)
                }
            }

            override fun onMessage(message: String?) {
                try {
                    val obj = JSONObject(message)
                    when (obj.optString("action")) {
                        "reload" -> {
                            handleReload(obj)
                        }
                        "clearBreakPoint" -> {
                            handleClearBreakPoint(obj)
                        }
                        "clearBreakPoints" -> {
                            handleClearBreakPoints(obj)
                        }
                        "setBreakPoint" -> {
                            handleSetBreakPoint(obj)
                        }
                        "continue" -> {
                            handleContinue()
                        }
                        "stop" -> {
                            handleStop()
                        }
                        "step" -> {
                            handleStep()
                        }
                        "eval" -> {
                            handleEval(obj)
                        }
                        else -> {
                            println(false)
                        }
                    }
                } catch (e: Exception) {}
            }

            override fun onError(ex: Exception?) {
                handler.postDelayed({
                    connectWithIP(IP, port, context)
                }, 1000)
            }

        }
    }

    private fun handleReload(obj: JSONObject) {
        obj.optString("source")?.let {
            val sourceFile = File(context?.cacheDir, System.currentTimeMillis().toString() + ".js")
            sourceFile.writeBytes(Base64.decode(it, 0))
            this.sourceURL = Uri.fromFile(sourceFile).toString()
            handler.post {
                this.delegate?.debuggerDidReload()
            }
        }
    }

    private fun handleClearBreakPoint(obj: JSONObject) {
        val bpIdentifier = obj.optString("path") + ":" + obj.optString("line")
        handler.post {
            try {
                (xtContext?.evaluateScript("window.XTDebug") as? V8Object)?.let {
                    XTContext.invokeMethod(it, "clearBreakpoint", listOf(bpIdentifier))
                }
            } catch (e: Exception) {}
        }
        this.activeBreakpoints.remove(bpIdentifier)
    }

    private fun handleClearBreakPoints(obj: JSONObject) {
        val path = obj.optString("path")
        handler.post {
            try {
                (xtContext?.evaluateScript("window.XTDebug") as? V8Object)?.let {
                    XTContext.invokeMethod(it, "clearBreakpoints", listOf(path))
                }
            } catch (e: Exception) {}
        }
        this.activeBreakpoints.filter { return@filter it.startsWith(path) }.forEach {
            this.activeBreakpoints.remove(it)
        }
    }

    private fun handleSetBreakPoint(obj: JSONObject) {
        val bpIdentifier = obj.optString("path") + ":" + obj.optString("line")
        handler.post {
            try {
                (xtContext?.evaluateScript("window.XTDebug") as? V8Object)?.let {
                    XTContext.invokeMethod(it, "setBreakpoint", listOf(bpIdentifier))
                }
            } catch (e: Exception) {}
        }
        this.activeBreakpoints.add(bpIdentifier)
    }

    private fun handleBreak(bpIdentifier: String, T: String, S: String) {
        this.sendBreak(bpIdentifier, T, S)
        this.breakpointLocking = true
        while (this.breakpointLocking) {
            try {
                Thread.sleep(100)
            } catch (e: Exception) { this.breakpointLocking = false }
        }
    }

    private fun handleContinue() {
        this.breakpointStepping = false
        this.breakpointLocking = false
    }

    private fun handleStep() {
        this.breakpointStepping = true
        this.breakpointLocking = false
        handler.postDelayed({
            this.breakpointStepping = false
        }, 100)
    }

    private fun handleStop() {
        this.breakpointStepping = false
        this.breakpointLocking = false
        handler.post {
            this.delegate?.debuggerDidTerminal()
        }
    }

    private fun sendBreak(bpIdentifier: String, T: String, S: String) {
        try {
            val obj = JSONObject()
            obj.put("type", "break")
            obj.put("bpIdentifier", bpIdentifier)
            obj.put("this", T)
            obj.put("scope", S)
            this.socket?.send(obj.toString())
        } catch (e: Exception) { e.printStackTrace() }
    }

    private fun handleEval(obj: JSONObject) {
        val expression = obj.optString("expression") ?: return
    }

    companion object {

        val sharedDebugger = XTDebug()

        fun debugWithIP(IP: String, port: String, context: Context) {
            sharedDebugger.connectWithIP(IP, port, context)
        }

    }

    class JSExports(val context: XTContext): XTComponentExport() {

        override val name: String = "_XTDebug"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "xtr_break", "xtr_break", arrayOf(String::class.java, String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_stepping", "xtr_stepping", arrayOf())
            return exports
        }

        fun xtr_break(bpIdentifier: String, T: String, S: String) {
            sharedDebugger.currentBpIdentifier = bpIdentifier
            if (sharedDebugger.breakpointStepping || sharedDebugger.activeBreakpoints.contains(bpIdentifier)) {
                sharedDebugger.handleBreak(bpIdentifier, T, S)
            }
        }

        fun xtr_stepping(): Boolean {
            return sharedDebugger.breakpointStepping
        }

    }

}